### PR TITLE
Fix building on RHEL 8.5

### PR DESCRIPTION
**Pull request by ThinkOpenly**

### CODE REVIEW

**Comment by mscastanho** [comment]
@ThinkOpenly The builds passed the `gcc_3` stage, but failed when creating the final distro packages. Example for Ubuntu:
```

dh_install
dh_missing --fail-missing
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libanl.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libthread_db.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libc.so.6 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/librt.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnss_dns.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/ld64.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnsl.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnss_files.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libm.so.6 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnss_db.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libBrokenLocale.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libutil.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libpthread.so.0 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnss_hesiod.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libnss_compat.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libresolv.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libdl.so.2 exists in debian/tmp but is
 not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power9/libcrypt.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libanl.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libthread_db.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libc.so.6 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/librt.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnss_dns.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/ld64.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnsl.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnss_files.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libm.so.6 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnss_db.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libBrokenLocale.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libutil.so.1 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libpthread.so.0 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnss_hesiod.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libnss_compat.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libresolv.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libdl.so.2 exists in debian/tmp but is not installed to anywhere
dh_missing: warning: /at_opt/at14.0-4-rc1/lib64/glibc-hwcaps/power10/libcrypt.so.1 exists in debian/tmp but is not installed to anywhere
```

And if you need to add further changes/commits to your PR, you can simply edit your branch locally and then use `git push <your-remote-name> hwcaps2 --force` that it will rewrite your PR (just be careful when you type it so you don't erase something else by mistake). No need to create a new one 😉

**Review by mscastanho — COMMENTED** [review]
@ThinkOpenly  IIUC, in the first cache update you make a copy of all libs already built by AT to `glibc-hwcaps`, and now `syslib-override` is just a "timestamp" file used by the second update to help you delete any old libs copied in the first update. Is that correct?

What I didn't get is why is this only being applied to `"$(AT_MAJOR_VERSION)" > 15`. Didn't we also need the `syslib-override` trick for AT 14/15 as well? I actually expected an AT 14 build on RHEL 8.5 to fail, but I tried and it completed fine, so I think I'm missing some important concept here.

**Comment by ThinkOpenly** [comment]
> @ThinkOpenly IIUC, in the first cache update you make a copy of all libs already built by AT to `glibc-hwcaps`, and now `syslib-override` is just a "timestamp" file used by the second update to help you delete any old libs copied in the first update. Is that correct?

Correct. (Maybe changing the semantics of `syslib-override` like that is a bit confusing. :confused: )

> What I didn't get is why is this only being applied to `"$(AT_MAJOR_VERSION)" > 15`. Didn't we also need the `syslib-override` trick for AT 14/15 as well? I actually expected an AT 14 build on RHEL 8.5 to fail, but I tried and it completed fine, so I think I'm missing some important concept here.

This is an excellent question. An early attempt at this fix just dumped everything in the `glibc-hwcaps` directory and left it there, but I (unsurprisingly) encountered packaging complaints about files that weren't accounted for. Since we hadn't needed those directories previously (for AT14 and previous), I just shut that down for those releases, and added the more careful cleanup step for AT15 (and later).  However, I tried to think about this more carefully...

There are a number of different scenarios.

| AT release | distro has hwcaps | run-time or build-time | AT glibc | |
| -- | -- | -- | -- | -- |
| AT14 and earlier | everywhere | run-time | 2.32 (no `glibc-hwcaps`) | There are no system `glibc-hwcaps`. AT optimized libraries supersede system libraries. |
| AT14 and earlier | everywhere | build-time | 2.32 (no `glibc-hwcaps`) | Search paths appear to exclude system "optimized" paths, but I'm *not sure why*. Could be getting lucky due to issue #2654. |
| AT15 and later | no (RHEL 8.4) | run-time | 2.34 (`glibc-hwcaps`) | There are no system `glibc-hwcaps`. AT optimized libraries supersede. |
| AT15 and later | yes (RHEL 8.5) | run-time | 2.34 (`glibc-hwcaps`) | System `glibc-hwcaps` supersede build, so we need to ship AT optimized libraries in `glibc-hwcaps`. |
| AT15 and later | no (RHEL 8.4) | build-time | 2.34 (`glibc-hwcaps`) | There are no system `glibc-hwcaps`, but system's optimized libraries override AT libraries until they are built, so we need to fake AT optimized libraries during the build. |
| AT15 and later | yes (RHEL 8.5) | build-time | 2.34 (`glibc-hwcaps`) | System's `glibc-hwcaps` libraries override AT libraries, so we need to fake AT `glibc-hwcaps` libraries until they are built. |

Of particular concern is AT14, build-time. Given the presence of optimized libraries on the system, it would seem that these should interfere with the build. They currently don't. I interrupted a build during the `gcc_3` stage (after the first `ldconfig` stage):
```
2022-01-27_17.15.30 Completed first dynamic loader cache update!                                                           
2022-01-27_17.15.30 Starting the build of gcc_2...                                                                       
2022-01-27_17.37.12 Completed the build of gcc_2!                                                                          
2022-01-27_17.37.12 Starting the build of gcc_3...                                                                  
^C
```
In the midst of this build, the AT libraries are used:
```
$ LD_DEBUG=libs ~/opt/at14.0-4-rc1/bin/gcc                                                                                                                                                        
   3570255:     find library=libm.so.6 [0]; searching
   3570255:      search path=/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/altivec/dfp:/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/altivec:/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/dfp:/home/pc/opt/at14.0-4-rc1/lib64/tls/power10:/home/pc/opt/at14.0-4-rc1/lib64/tls/altivec/dfp:/home/pc/opt/at14.0-4-rc1/lib64/tls/altivec:/home/pc/opt/at14.0-4-rc1/lib64/tls/dfp:/home/pc/opt/at14.0-4-rc1/lib64/tls:/home/pc/opt/at14.0-4-rc1/lib64/power10/altivec/dfp:/home/pc/opt/at14.0-4-rc1/lib64/power10/altivec:/home/pc/opt/at14.0-4-rc1/lib64/power10/dfp:/home/pc/opt/at14.0-4-rc1/lib64/power10:/home/pc/opt/at14.0-4-rc1/lib64/altivec/dfp:/home/pc/opt/at14.0-4-rc1/lib64/altivec:/home/pc/opt/at14.0-4-rc1/lib64/dfp:/home/pc/opt/at14.0-4-rc1/lib64             (system search path)
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/altivec/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/altivec/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/power10/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/altivec/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/altivec/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/tls/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/power10/altivec/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/power10/altivec/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/power10/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/power10/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/altivec/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/altivec/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/dfp/libm.so.6
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/libm.so.6
   3570255:
   3570255:     find library=libc.so.6 [0]; searching
   3570255:      search path=/home/pc/opt/at14.0-4-rc1/lib64            (system search path)
   3570255:       trying file=/home/pc/opt/at14.0-4-rc1/lib64/libc.so.6
```
Although, do note that these libraries are found via "search", not via the loader cache. The loader cache display is troubling:
```
$ ~/opt/at14.0-4-rc1/sbin/ldconfig -p | grep lib[cm]\\.
        libm.so.6 (libc6,64bit, hwcap: 0x0000400000000000, OS ABI: Linux 3.10.0) => /lib64/power9/libm.so.6
        libm.so.6 (libc6,64bit, OS ABI: Linux 4.18.0) => /home/pc/opt/at14.0-4-rc1/lib64/libm.so.6
        libm.so.6 (libc6,64bit, OS ABI: Linux 3.10.0) => /lib64/libm.so.6
        libm.so (libc6,64bit, OS ABI: Linux 4.18.0) => /home/pc/opt/at14.0-4-rc1/lib64/libm.so
        libm.so (libc6,64bit, OS ABI: Linux 3.10.0) => /lib64/libm.so
[--]
        libc.so.6 (libc6,64bit, hwcap: 0x0000400000000000, OS ABI: Linux 3.10.0) => /lib64/power9/libc.so.6
        libc.so.6 (libc6,64bit, OS ABI: Linux 4.18.0) => /home/pc/opt/at14.0-4-rc1/lib64/libc.so.6
        libc.so.6 (libc6,64bit, OS ABI: Linux 3.10.0) => /lib64/libc.so.6
```
That the loader cache is not being used has been reported in issue #2654.
It could be that if the loader cache was being used, there would be an issue building AT14. The correct fix here, then, is probably to "fake" the optimized AT libraries in some way that they can be similarly cleaned up before packaging. Testing may have to wait until #2654 is fixed.

**Comment by tuliom** [comment]
@ThinkOpenly I noticed [4 patches like this one merged in master](https://github.com/advancetoolchain/advance-toolchain/commit/ac16e245312758bd8861340db2d77821c6832d24).
Are they related?

**Comment by ThinkOpenly** [comment]
> @ThinkOpenly I noticed [4 patches like this one merged in master](https://github.com/advancetoolchain/advance-toolchain/commit/ac16e245312758bd8861340db2d77821c6832d24). Are they related?

This branch is based on that (now merged) branch. I believe those 4 commits will disappear when this branch is rebased.

**Comment by tuliom** [comment]
>  I believe those 4 commits will disappear when this branch is rebased.

@ThinkOpenly I don't think that's how git works, i.e. git won't change the commit history.
For this PR to be merged, it has to be based on the code that is already on master. It must include only the new changes. Otherwise, the merge button will be disabled and we won't be able to merge it.

However, iff the 4 patches now on master were merged by mistake, it's fair to create a PR reverting them, i.e. with new commits created by `git-revert`.  This would allow to merge these 4 patches later without having to change them.

**Comment by mscastanho** [comment]
@tuliom there were 2 parallel PRs, both relying on these 4 commits. The first (#2635) has already been merged without changes to those commits. So a simple rebase should work. Doing it locally throws some warnings, but works, not sure why GitHub is complaining:
```
$ git rebase master
warning: skipped previously applied commit 1e90796c
warning: skipped previously applied commit 8c0a6726
warning: skipped previously applied commit 1b594224
warning: skipped previously applied commit bbdce6c8
hint: use --reapply-cherry-picks to include skipped commits
hint: Disable this message with "git config advice.skippedCherryPicks false"
Successfully rebased and updated refs/heads/hwcaps2.

$ git log --oneline -n2
bdaa8a21 (HEAD -> hwcaps2) Fix build when system has glibc-hwcaps libraries
616ac0c8 (tag: at14.0-4-rc2, upstream/master, upstream/HEAD, master) Prepare for AT 14.0-4 RC2
```

**Comment by tuliom** [comment]
@mscastanho Github does not rebase if there are conflicts.

File: scripts/helpers/utilities.mk
Comment (tuliom) [inline scripts/helpers/utilities.mk:400]: OK. This is replacing the syslib-override mechanism with glibc-hwcaps for all AT versions supported at the moment.

File: scripts/helpers/utilities.mk
Comment (tuliom) [inline scripts/helpers/utilities.mk:446]: Why is this code restricted to `AT > 15`?
What happens on AT 13, 14 and 15? Don't they need `syslib-override`?

File: scripts/helpers/utilities.mk
Comment (tuliom) [inline scripts/helpers/utilities.mk:461]: This is turning syslib-override into a file. What does this file do?
Is it being used just to indicate which files should be removed?

**Review by tuliom — CHANGES_REQUESTED** [review]
I have a few of questions about the impact of this change to AT <= 15.

Besides that, I think we need a manual rebase on top of `master` before we can merge this PR.

**Comment by ThinkOpenly** [comment]
> I think we need a manual rebase on top of `master` before we can merge this PR.

I'll do that. I wanted to get it into a pull request before the "hwcaps" branch was merged, so that the automatic builds would build the world. Now that it is merged, and all the checks have completed, I'll rebase and force-push. I guess we'll rebuild the world again. :disappointed:

File: scripts/helpers/utilities.mk
Comment (ThinkOpenly) [reply scripts/helpers/utilities.mk:446]: I confess to some confusion on my part here. When I did not so restrict, the builds failed. It may have been my sloppiness in cleaning up after faking the `syslib-override` resulting in files in the build tree which were not packaged, and the requisite errors.  That being said, the builds now all work.  The earlier AT releases may be benefiting from using `DT_RUNPATH` and not using the loader cache, though. The right thing to do is probably maintain the `syslib-override` behavior (likely renamed `glibc-hwcaps` for consistency with modern glibc).

File: scripts/helpers/utilities.mk
Comment (ThinkOpenly) [reply scripts/helpers/utilities.mk:461]: > This is turning syslib-override into a file. What does this file do? Is it being used just to indicate which files should be removed?

Correct. Anything with a timestamp _before_ the timestamp of this file is a fake, and needs to be removed. Not every file has an optimized version, so keeping track explicitly is error-prone. This is a fairly simple shortcut.

**Review by tuliom — APPROVED** [review]

**Pull request merged**
